controller/src/test/java/io/pravega/controller/store/stream/ZkStreamTest.java

```diff
@@ -580,6 +581,57 @@ public void testGetActiveTxn() throws Exception {
         assertEquals(1, result.size());
     }
 
+    @Test(timeout = 10000)
+    public void testStreamRecreation() {
+        // We will first create stream. Verify that its metadata is present in the cache.  
+        ZKStoreHelper storeHelper = new ZKStoreHelper(cli, executor);
+        String scope = "scope";
+        String stream1 = "streamToDelete";
+        ZKStream stream = new ZKStream(scope, stream1, storeHelper);
+        final int startingSegmentNumber = 0;
+        storeHelper.createZNodeIfNotExist("/store/scope").join();
+        final ScalingPolicy policy1 = ScalingPolicy.fixed(2);
+        final StreamConfiguration configuration1 = StreamConfiguration.builder().scalingPolicy(policy1).build();
+        stream.create(configuration1, System.currentTimeMillis(), startingSegmentNumber).join();
+        stream.createStreamPositionNodeIfAbsent(0).join();
+        stream.updateState(State.ACTIVE).join();
+        Long creationTime = stream.getCreationTime().join();
+        Integer position = stream.getStreamPosition().join();
+        assertEquals(0, position.intValue());
+        ZKStoreHelper.ZkCacheKey<Integer> key = new ZKStoreHelper.ZkCacheKey<>(stream.getCreationPath(), position.toString(), x -> BitConverter.readInt(x, 0));
+        VersionedMetadata<?> cachedCreationTime = storeHelper.getCache().getCachedData(key).join();
+        // verify that both times are same
+        assertEquals(creationTime, cachedCreationTime.getObject());
+        // delete stream.
+        stream.updateState(State.SEALING).join();
+        stream.updateState(State.SEALED).join();
+        stream.deleteStream().join();
+
+        // refresh the stream object to indicate new request context
+        stream.refresh();
+        // verify that metadata doesnt exist in the store.
+        AssertExtensions.assertFutureThrows("Stream deleted", stream.getCreationTime(), e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException);
+        
+        // verify that cached entries still exist. 
+        VersionedMetadata<?> cachedCreationTimeExists = storeHelper.getCache().getCachedData(key).join();
+        assertEquals(cachedCreationTime.getObject(), cachedCreationTimeExists.getObject());
+        
+        // create stream again.
+        stream.create(configuration1, System.currentTimeMillis(), startingSegmentNumber).join();
+        stream.createStreamPositionNodeIfAbsent(1).join();
+        stream.updateState(State.ACTIVE).join();
+
+        Long creationTimeNew = stream.getCreationTime().join();
+        Integer positionNew = stream.getStreamPosition().join();
+        assertEquals(1, positionNew.intValue());
+
+        ZKStoreHelper.ZkCacheKey<Integer> keyNew = new ZKStoreHelper.ZkCacheKey<>(stream.getCreationPath(), positionNew.toString(), x -> BitConverter.readInt(x, 0));
+        VersionedMetadata<?> cachedCreationTimeNew = storeHelper.getCache().getCachedData(keyNew).join();
+        // verify that both times are different
```

Comment:
This is a good test case. What about the case in which the stream is deleted and the cached entry still exists? Will the cached entry for the deleted stream be visible to a request?

Reply:
when a stream is deleted, the controller instance where the delete took place cleans up the cache.. but the entry may have been cached on other controller instances that are unaware of the delete.. 
the stream has a unique 'id' upon each new creation.. we use this `id` while caching.. so any cache for an old stream is not visible to anyone requesting data for a new stream with a different id

Reply:
I was not actually asking you to explain the case, but about the presence of a test case for it, but since you got into it, let's talk about it:

> any cache for an old stream is not visible to anyone requesting data for a new stream with a different id

I'm not sure I understand. A client gets the id from the controller, and given that we can have stale cache entries in other controller instances, the client could get a stale id for the same stream name, no?

Reply:
Clients deal with stream names only. 
The id is something controllers use internally for caching only for different generations of streams with same name. 
The metadata is stored at a path that only depends on the stream name and nothing changes there.

Reply:
We had to resolve this issue offline, it is not trivial to understand why this is correct.